src/libpmemfile/fd_pool.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2016, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE
+
+#include "fd_pool.h"
+#include "libsyscall_intercept_hook_point.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <fcntl.h>
+#include <syscall.h>
+#include <sys/mman.h>
+#include <asm-generic/errno.h>
+#include <stdlib.h>
+
+
+/*
+ * O(1) lookup: is fd allocated for the fd pool, i.e.:
+ * it is either associated with a pmemfile, or just waiting in the pool.
+ * Such file descriptors should not be handled by the kernel.
+ */
+static bool is_fd_allocated[PMEMFILE_MAX_FD + 1];
+
+/*
+ * A stack of file descriptors in the pool, not yet used for pmemfile.
+ * This allows us to fetch a new fd in O(1), or to put an fd back into
+ * the pool in O(1)
+ */
+static long fd_pool[PMEMFILE_MAX_FD + 1];
+
+/*
+ * allocated_count - how many file descriptors are fetch from the kernel with
+ * open("/dev/null", O_RDONLY)
+ *
+ * Upon initialization, the fd_pool array contains all these fds.
+ *
+ * TODO: do we actually need to keep track of this?
+ */
+static size_t allocated_count;
+
+/*
+ * next_item_to_use - the index pointing to an element in the fd_pool array.
+ * It is initialized to allocated_count - 1, and decreased each time an item is
+ * used. Increased every time a previously used item item is placed back
+ * into the pool. If this becomes negative, it means the pool is empty i.e.:
```

Comment:
Oh, I was planning to get rid of the pool, and just open ```/dev/null``` for each new fd. That is a far more flexible approach. I'm not sure we should rather implement that before merging, or try to patch up the current implementation according to review.